### ISSUE TITLE
appium should use an already-running iOS simulator if appropriate

### ISSUE THREAD

**Issue by jlipps** [appium/appium#6385]
iOS sims take forever to boot. If one is running, and if there is no reason to close it (e.g., to support a desired cap which requires rewriting of plists, or because the requested simulator is of a different type or udid than the running one, or because we've requested a full reset), Appium should simply use the running simulator.

**Comment by jlipps** [comment]
@imurchie I ran a quick test and it seems we currently kill all current sims before every session regardless of whether that is absolutely necessary. Seems like something we can/should optimize given the boot time of the sim

**Comment by Simon-Kaz** [comment]
I played around with this a bit before. I remember that when i set the:

capabilities.setCapability("noReset", true);

The simulator would stay and be reused between tests. I don't remember if the app was installed over the existing one, I believe it was reused for all tests? 

Here's my code, I was looking for ways to speed up test execution on Sauce Labs.

https://github.com/Simon-Kaz/AppiumSauce/blob/master/src/test/java/app/nativeApp/ios/NativeIOSTest.java

That was all done on Appium 1.4.16 btw, I'm not sure if 1.15 broke that.

**Comment by jlipps** [comment]
I do think it changed a bit in 1.5 with our new sim killing logic

**Comment by SrinivasanTarget** [comment]
https://github.com/appium/appium/issues/5208

**Comment by imurchie** [comment]
The current code does kill all the simulators all the time. Indeed, in all the released versions it kills them even on real device tests. Alas.

**Comment by saikrishna321** [comment]
https://github.com/appium/appium/issues/5768

**Comment by artgo** [comment]
+1 Same issue

**Comment by hiddentao** [comment]
Hitting this issue now, and the simulator doesn't finish booting quickly enough for tests to then get run.

**Comment by imurchie** [comment]
The system ought to wait for the sim to boot. Do you have logs, @hiddentao, for this happening?

**Comment by hiddentao** [comment]
This is what I see:

```
[debug] [BaseDriver] Unzipping /var/folders/4v/br6x6mlx113235v1lz39nwfc0000gn/T/20161017-52513-eysr0b.gvzfwp14i/appium-app.zip
[debug] [BaseDriver] Testing zip archive: /var/folders/4v/br6x6mlx113235v1lz39nwfc0000gn/T/20161017-52513-eysr0b.gvzfwp14i/appium-app.zip
[BaseDriver] Unzipped local app to '/var/folders/4v/br6x6mlx113235v1lz39nwfc0000gn/T/20161017-52513-eysr0b.gvzfwp14i/Payload/myapp.app'
[debug] [XCUITest] Checking whether app '/var/folders/4v/br6x6mlx113235v1lz39nwfc0000gn/T/20161017-52513-eysr0b.gvzfwp14i/Payload/myapp.app' is actually present
[debug] [XCUITest] App is present
[debug] [ios-app-utils] Getting bundle ID from app '/var/folders/4v/br6x6mlx113235v1lz39nwfc0000gn/T/20161017-52513-eysr0b.gvzfwp14i/Payload/myapp.app': 'com.prod.app'
[debug] [iOSLog] Starting iOS 10.1 simulator log capture
[debug] [iOSLog] System log path: /Users/ram/Library/Logs/CoreSimulator/C6F7B300-21AA-4C3A-9AC8-49CD36FB85A2/system.log
[XCUITest] Setting up simulator
[debug] [iOS] No reason to set locale
[debug] [iOS] No iOS / app preferences to set
[XCUITest] Simulator with udid 'C6F7B300-21AA-4C3A-9AC8-49CD36FB85A2' not booted. Booting up now
[debug] [iOSSim] Killing all iOS Simulators
[iOSSim] Starting simulator with command: open /Applications/Xcode.app/Contents/Developer/Applications/Simulator.app --args -CurrentDeviceUDID C6F7B300-21AA-4C3A-9AC8-49CD36FB85A2
[iOSSim] Simulator log at '/Users/ram/Library/Logs/CoreSimulator/C6F7B300-21AA-4C3A-9AC8-49CD36FB85A2/system.log'
[iOSSim] Tailing simulator logs until we encounter the string "com.apple.springboard"
[iOSSim] We will time out after 60000ms
[debug] [iOSSim] Waiting an extra 10000ms for the simulator to really finish booting
[debug] [iOSSim] Done waiting extra time for simulator
[iOSSim] Simulator booted in 26773ms
```

The simulator was still booting (apple logo with progress bar) when the logs showed this.

**Comment by imurchie** [comment]
Hmm. Ok. That means the boot indicator is not correct. Appium waits for something particular to be written to the logs to know that the sim is booted, and Apple helpfully removed the very reliable line in iOS 10+.

**Comment by lanzelot1989** [comment]
Hi, is there any progress on this?

I am currently facing this issue. I am trying to run several simulators and tests in parallel on one Mac machine by distributing the tests through Selenium Grid.
I am using fbsimctl to start the simulators and then start up appium (one for each simulator). Then after I send a test case to the grid, all but one simulators are killed.

Trying to run several simulators and only starting one Appium server showed that it also made no difference whether the simulator was connected to the grid or not. All other simulators were killed nevertheless.

Setting "noReset" to true does not seem to have any effect.

## Environment
XCode 8.2.1
Appium 1.6.3
Mac 10.12.2

**Comment by dimagtf** [comment]
still no updates?

**Comment by mykola-mokhnach** [comment]
@dimagtf Try 1.6.4-beta. It should be already implemented there

**Comment by dimagtf** [comment]
@mykola-mokhnach, thanks, but this issue was not fixing in 1.6.4-beta. 
Appium continues kill other open simulators.

**Comment by mykola-mokhnach** [comment]
it's just does not restart the simulator if this is not needed, which saves time between session restart. Regarding multiple simulators - we don't support it now and this task is not present in the backlog. Probably, we might think about this after simctl is replaced with fbsimctl and the other system related stuff is resolved. But this will require to rewrite a lot of code (and tests) and not gonna happen in the close perspective.

**Comment by dimagtf** [comment]
It means, that does not any way to test on multiple ios simulators via appium now and in the near future?

**Comment by lanzelot1989** [comment]
@mykola-mokhnach It would be a well appreciated feature for us as it currently only leaves us the option to scale via multiple macOS machines (which means that there is *a lot* of unnecessary overhead) if I understood the facts correctly.

I am looking forward to this, because as you indicated, running multiple simulators is not a hassle any more with fbsimctl.

What would it need for you to introduce this task to the backlog?

**Comment by mmcintyre123** [comment]
I would also like this feature.  Should be very easy to do.

**Comment by dmshantanu** [comment]
Is the support for multiple ios devices available now?  Need it for sure.

**Comment by alter-al** [comment]
> Is the support for multiple ios simulators available now?
not available now

This is very strange, because other frameworks yet use fbsimctl for quite long time. Why is it so difficult to implement for an appium?

**Comment by saikrishna321** [comment]
@jlipps @mykola-mokhnach  by setting noReset cap the simulator does not close, is there any other reason this issue is still open ?

**Comment by mykola-mokhnach** [comment]
@saikrishna321 yep, it can be closed now

**Comment by alter-al** [comment]
Sorry, but in Appium 1.6.5 with caps: "noReset" I still can not start more one simulator.
Appium trying to start test at all running simulators and kills them all, except one.

Was "noReset" supposed to help, not FbSim??

**Comment by saikrishna321** [comment]
@alter-al "noReset" will run the tests in open simulator and won't restart a new sim on every test

**Comment by alter-al** [comment]
@saikrishna321, but it is not resolving this [issue](https://github.com/appium/appium/issues/5768), am I wrong?

**Comment by lanzelot1989** [comment]
Is there currently an issue that encapsulates the need for more than one simulator per macOS machine?
Because this was described in more detail here, otherwise I'd be happy to create a new one.

**Comment by saikrishna321** [comment]
@alter-al running on multiple simulators is completely a different case

**Comment by mykola-mokhnach** [comment]
@alter-al We are currently working on parallel simulator testing support starting with Xcode9. The most recent changes are available in Appium 1.6.6-beta
It is very welcome to test it and provide constructive feedback, so we can fix all the major issues before release.

The example of parallel test: https://github.com/appium/appium-xcuitest-driver/blob/master/test/functional/parallel/parallel-simulators-e2e-specs.js
